test/harry/main/org/apache/cassandra/harry/ColumnSpec.java

```diff
@@ -0,0 +1,605 @@
+    {
+        return name + '(' + type.toString() + ")";
+    }
+
+    public Generator<T> gen()
+    {
+        return gen;
+    }
+
+    public static <T> ColumnSpec<T> pk(String name, DataType<T> type, Generator<T> gen)
+    {
+        return new ColumnSpec<>(name, type, gen, Kind.PARTITION_KEY);
+    }
+
+    public static <T> ColumnSpec<T> pk(String name, DataType<T> type)
+    {
+        return new ColumnSpec<>(name, type, (Generator<T>) TypeAdapters.forValues(type.asServerType()), Kind.PARTITION_KEY);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T> ColumnSpec<T> ck(String name, DataType<T> type, Generator<T> gen, boolean isReversed)
+    {
+        return new ColumnSpec(name, isReversed ? ReversedType.getInstance(type) : type, gen, Kind.CLUSTERING);
+    }
+
+    public static <T> ColumnSpec<T> ck(String name, DataType<T> type, boolean isReversed)
+    {
+        return new ColumnSpec(name, isReversed ? ReversedType.getInstance(type) : type,
+                              TypeAdapters.forValues(type.asServerType()),
+                              Kind.CLUSTERING);
+    }
+
+
+    public static <T> ColumnSpec<T> regularColumn(String name, DataType<T> type, Generator<T> gen)
+    {
+        return new ColumnSpec<>(name, type, gen, Kind.REGULAR);
+    }
+
+    public static <T> ColumnSpec<T> regularColumn(String name, DataType<T> type)
+    {
+        return new ColumnSpec(name, type, TypeAdapters.forValues(type.asServerType()), Kind.REGULAR);
+    }
+
+    public static <T> ColumnSpec<T> staticColumn(String name, DataType<T> type, Generator<T> gen)
+    {
+        return new ColumnSpec<>(name, type, gen, Kind.STATIC);
+    }
+
+    public static <T> ColumnSpec<T> staticColumn(String name, DataType<T> type)
+    {
+        return new ColumnSpec(name, type, TypeAdapters.forValues(type.asServerType()), Kind.STATIC);
+    }
+
+    public enum Kind
+    {
+        CLUSTERING, REGULAR, STATIC, PARTITION_KEY
+    }
+
+    public static abstract class DataType<T>
+    {
+        protected final String cqlName;
+
+        protected DataType(String cqlName)
+        {
+            this.cqlName = cqlName;
+        }
+
+        public abstract /* unsigned */ long typeEntropy();
+
+        public boolean isReversed()
+        {
+            return false;
+        }
+
+        public abstract AbstractType<?> asServerType();
+
+        public final String toString()
+        {
+            return cqlName;
+        }
+
+        public final boolean equals(Object o)
+        {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            DataType<?> dataType = (DataType<?>) o;
+            return Objects.equals(cqlName, dataType.cqlName);
+        }
+
+        public final int hashCode()
+        {
+            return Objects.hash(cqlName);
+        }
+
+        public Comparator<T> comparator()
+        {
+            return (o1, o2) -> ((Comparable) o1).compareTo(o2);
```

Comment:
I have a recency bias here, but there were 2-3 bugs found in `AbstractType` due to this pattern... when someone extended the type they didn't know that they *should* override random methods, which then caused the type to be corrupted or buggy... 

In `TypeSupport` its explicit for similar reasons, when you are defining the type you are the only one that knows what java type is used (client/server have different types) so you *must* think about ordering at that moment... extension allows you to avoid thinking about this which leads to bugs.

Reply:
Made the change you suggested.